Heroku calls

nellshamrell opened the pull request:
This adds in the ability to get a list of apps associated with a Heroku account. I will be adding in several more Heroku specific commands in the next pull request.

smarnach commented on src/authorizations/users.rs:
```diff
@@ -1,26 +1,39 @@
-        .split(',')
-        .map(String::from)
-        .collect()
+    let bot_config = Config::default();
```
I prefer to generate the configuration once in `main()` and then pass it around to whereever it is needed. Doing this reduces the dependency on global state, and makes testing a lot easier in my experience. Many functions currently depending on environment variables will become pure functions this way.

pietroalbini commented on src/commands/heroku.rs:
```diff
@@ -0,0 +1,66 @@
+
+#[command]
+pub fn get_apps(ctx: &mut Context, msg: &Message, _args: Args) -> CommandResult {
+    let response = heroku_client().get().apps().execute::<Value>();
```
Is there a reason why you're getting a `Value` instead of a `Vec<HerokuApp>`? If you add `#[derive(Deserialize)]` to `HerokuApp` you can just use:

```suggestion
    let response = heroku_client().get().apps().execute::<Vec<HerokuApp>>();
```

...and remove all the custom deserialization code.

nellshamrell replied:
Sounds great - adding that in!

nellshamrell replied:
Done!

pietroalbini commented on src/commands/heroku.rs:
```diff
@@ -0,0 +1,66 @@
+
+fn heroku_client() -> heroku_rs::client::Heroku {
+    let heroku_api_key = Config::default().heroku_api_key;
+    Heroku::new(heroku_api_key).unwrap()
```
The `Heroku` struct embeds a HTTP client, so if we reuse it across requests instead of creating a new one each command we could rely on keep-alive connections to get a performance boost.

nellshamrell replied:
I experimented with creating a heroku client in the run() method in lib.rs. I wanted to pass the client to each of the commands. Unfortunately, the only way to do this is to write to the data attribute on the Serenity Client. In order to pass an instance of a struct to the data attribute, it needs to implement the TypeMapKey trait (which is defined in Serenity). The Heroku struct is defined in the Heroku-rs crate...so unfortunately I cannot implement the TypeMapKey trait on the Heroku struct as it is right now. I'm thinking about leaving it as creating a Heroku client for each command for now, but if we are having performance issues we can certainly revisit this.

pietroalbini replied:
I don't think implementing `TypeMapKey` for the `Heroku` struct is needed, as the key doesn't need to be the same type as the value. You could do something like this:

```rust
pub(crate) struct HerokuClient;

impl TypeMapKey for HerokuClient {
    type Value = heroku_rs::client::Client;
}
```

...and then use the `HerokuClient` key to retrieve the proper client.

smarnach replied:
I wouldn't be particularly concerned with performance for the bot – it will probably never get more than one request per minute or so. However, creating the Heroku client creates an event loop, a thread pool and a reactor per thread. Doing that for each request just _feels_ a bit too much. :)

(I'm also a bit surprised that the client creates the event loop internally. If it needs one, I'd expect it to accept the event loop as a parameter.)

nellshamrell replied:
I've been attempting this with:
```rust
use heroku_rs::client::{Heroku};

pub(crate) struct HerokuClient;

impl TypeMapKey for HerokuClient {
    type Value = heroku_rs::client::Heroku;
}
```

and 

```rust
 data.insert::<HerokuClient>(heroku_client);
```

However, this is returning a number of errors:

```bash
error[E0277]: `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>` cannot be sent between threads safely
  --> src/lib.rs:47:14
   |
47 |         data.insert::<HerokuClient>(heroku_client);
   |              ^^^^^^ `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>` cannot be sent between threads safely
   |
   = help: within `heroku_rs::client::Heroku`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>`
   = note: required because it appears within the type `heroku_rs::client::Heroku`
   = note: required because of the requirements on the impl of `typemap::internals::Implements<(dyn unsafe_any::UnsafeAny + std::marker::Send + std::marker::Sync + 'static)>` for `heroku_rs::client::Heroku`

error[E0277]: `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>` cannot be sent between threads safely
  --> src/lib.rs:47:14
   |
47 |         data.insert::<HerokuClient>(heroku_client);
   |              ^^^^^^ `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>` cannot be sent between threads safely
   |
   = help: within `heroku_rs::client::Heroku`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>`
   = note: required because it appears within the type `heroku_rs::client::Heroku`
   = note: required because of the requirements on the impl of `typemap::internals::Implements<(dyn unsafe_any::UnsafeAny + std::marker::Send + std::marker::Sync + 'static)>` for `heroku_rs::client::Heroku`

error[E0277]: `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>` cannot be shared between threads safely
  --> src/lib.rs:47:14
   |
47 |         data.insert::<HerokuClient>(heroku_client);
   |              ^^^^^^ `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>` cannot be shared between threads safely
   |
   = help: within `heroku_rs::client::Heroku`, the trait `std::marker::Sync` is not implemented for `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>`
   = note: required because it appears within the type `heroku_rs::client::Heroku`
   = note: required because of the requirements on the impl of `typemap::internals::Implements<(dyn unsafe_any::UnsafeAny + std::marker::Send + std::marker::Sync + 'static)>` for `heroku_rs::client::Heroku`

error[E0277]: `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>` cannot be shared between threads safely
  --> src/lib.rs:47:14
   |
47 |         data.insert::<HerokuClient>(heroku_client);
   |              ^^^^^^ `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>` cannot be shared between threads safely
   |
   = help: within `heroku_rs::client::Heroku`, the trait `std::marker::Sync` is not implemented for `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>`
   = note: required because it appears within the type `heroku_rs::client::Heroku`
   = note: required because of the requirements on the impl of `typemap::internals::Implements<(dyn unsafe_any::UnsafeAny + std::marker::Send + std::marker::Sync + 'static)>` for `heroku_rs::client::Heroku`

error: aborting due to 4 previous errors
```

This appears to be because of how the fields are defined for [the Heroku struct in heroku-rs](https://github.com/bensadiku/heroku-rs/blob/d3e97651c840d2cbf5d7892619329bfa0b908340/src/client.rs#L33)

```rust
#[derive(Debug)]
pub struct Heroku {
    token: String,
    core: Rc<RefCell<Core>>,
    client: Rc<Client<HttpsConnector>>,
}
```

Rc<T> and RefCell<T> are, to my knowledge, not shareable across threads.

I will be looking into contributing to heroku-rs, to see if I can make the client safe to share across threads, but it feels a bit out of scope for this pull request.  Once this is merged, I am happy to create an issue to figure out how to share the Heroku client across threads, including contributing to heroku-rs.

nellshamrell replied:
Issue has been created! https://github.com/rust-lang/crates-io-ops-bot/issues/19

pietroalbini replied:
Sounds good.

nellshamrell commented:
@smarnach @pietroalbini - thank you so much for the feedback, I've done quite a bit of refactoring to incorporate it in :)

nellshamrell commented:
@smarnach @pietroalbini - thank you so much for the feedback, I've done quite a bit of refactoring to incorporate it in :)

smarnach commented:
One solution for the `Config` object is to wrap it in an `Arc`. This allows you to sneak a shared reference out without holding on to the read lock for the context – see https://github.com/smarnach/crates-io-ops-bot/commit/30dda8074a3c70dd0ee7b6e38a976f6b66cfea40 for a quick implementation attempt. This also makes it easy to wrap the config extraction in a function.

Why do you have a `config` module nested inside the `config` module, by the way? I think one level of `config` would feel enough. :)

pietroalbini commented on src/commands/heroku.rs:
```diff
@@ -0,0 +1,57 @@
+    match response {
+        Ok((_headers, _status, json)) => {
+            if let Some(mut json) = json {
+                processed_app_list.append(&mut json);
```
Is there a reason why you don't just use the `Vec` in the `json` variable instead of moving its elements to another one?

nellshamrell replied:
I was having some trouble with persisting the json variable  out of the loop (and then passing it to msg.reply). Going to see if I can find a more elegant solution.

nellshamrell replied:
I just did a small refactor where I

* [created the processed_app_list variable outside of the match and if expression](https://github.com/rust-lang/crates-io-ops-bot/blob/ad0c92d1672af3c015f49ac4e82776ebd8407ba5/src/commands/heroku.rs#L28)
* within the match and if expression, [assign the processed_list_variable to the returned json](https://github.com/rust-lang/crates-io-ops-bot/blob/ad0c92d1672af3c015f49ac4e82776ebd8407ba5/src/commands/heroku.rs#L33)
* outside of the match and if expression [pass the processed_app_list to the app_response function, then pass the result of that function to the msg.reply funtion](https://github.com/rust-lang/crates-io-ops-bot/blob/ad0c92d1672af3c015f49ac4e82776ebd8407ba5/src/commands/heroku.rs#L40)

Is there a better way to do this? I'm completely open to refining it further!

pietroalbini replied:
Personally I'd replace the match and the reply with:

```rust
msg.reply(ctx, match response {
    Ok((_, _, Some(apps))) => app_response(apps),
    Ok((_, _, None)) => "You have no Heroku apps".into(),
    Err(err) => {
        println!("Err {}", err);
        "An error occured while fetching your Heroku apps".into()
    }
})?;

pietroalbini commented on src/authorizations/users.rs:
```diff
@@ -1,29 +1,33 @@
+
+fn authorized_users(config: &Config) -> Vec<String> {
+    let split_string = config.authorized_users.split(',');
+    let auth_users: Vec<String> = split_string.map(String::from).collect();
```
Instead of storing the list of users as a string in the config and parsing it every time, it would be better to parse it beforehand and store it in a `HashSet` in the config.

nellshamrell replied:
Excellent point - done!

nellshamrell commented:
> Why do you have a config module nested inside the config module, by the way? I think one level of config would feel enough. :)

I'm...not entirely sure. I blame it on lack of caffeine. Fixed!

nellshamrell commented:
> One solution for the Config object is to wrap it in an Arc. This allows you to sneak a shared reference out without holding on to the read lock for the context – see smarnach@30dda80 for a quick implementation attempt. This also makes it easy to wrap the config extraction in a function.

Excellent point - done!

pietroalbini commented on src/authorizations/users.rs:
```diff
@@ -1,48 +1,27 @@
-pub fn is_authorized(id: &str) -> bool {
-   authorized_users().iter().any(|i| i == id)
+pub fn is_authorized(id: &str, config: &Config) -> bool {
+    config.authorized_users.iter().any(|i| i == id)
```
Now that this is an `HashSet` you can simply use the contains method.

nellshamrell replied:
ah! excellent point!

nellshamrell commented:
@pietroalbini @smarnach - thank you so much for the very thoughtful review and feedback! This is Open Source at its finest!

I believe I have addressed all of your concerns - please let me know if this is good to merge!

smarnach commented:
Regarding the Heroku client library, it looks like that library is not a good fit for us. It probably causes more problems than it solves.

I haven't looked at the code in detail, but it's completely unclear to me why a REST API client should own a Tokio event loop with all that entails. The first version of the lib was released less than a week ago, and it hardly has any documentation. It's very well possible that it is meant for a very specific use case, and it looks like that use case is quite different from ours.

From what I've seen, I believe it would be way less trouble to simply send the REST request with `reqwest` than trying to twist `heroku_rs` to fit our use case.

nellshamrell commented:
@smarnach We certainly could use reqwest to send the API commands and I have that under consideration. My goal with using heroku_rs was to work with the creator (whom I have been in contact with) of it to improve it with a real world use scenario. If it is truly not fitting our needs, I am happy to send the requests using reqwest, but I want to try to work with heroku_rs for a little longer.

smarnach commented:
@nellshamrell Thanks for your reply. It's good to know you are in contact with the creator – it would indeed be nice to turn the library into something more generally useful.
